File: dirty-flag/src/main/java/com/iluwatar/dirtyflag/World.java
package com.iluwatar.dirtyflag;

import java.util.ArrayList;
import java.util.List;

/**
 * 
 * A middle-layer app that calls/passes along data from the back-end.
 * 
 * @author swaisuan
 *
 */
public class World {

  private List<String> countries;
  private DataFetcher df;

  public World() {
    this.countries = new ArrayList<String>();
    this.df = new DataFetcher();
  }

  /**
   * 
   * Calls {@link DataFetcher} to fetch data from back-end.
   * 
   * @return List of strings
   */
  public List<String> fetch() {
    List<String> data = df.fetch();

    countries = data.isEmpty() ? countries : data;

    return countries;
  }
}

File: dirty-flag/src/main/kotlin/com/iluwatar/dirtyflag/World.kt
/*
 * This project is licensed under the MIT license. Module model-view-viewmodel is using ZK framework licensed under LGPL (see lgpl-3.0.txt).
 *
 * The MIT License
 * Copyright © 2014-2022 Ilkka Seppälä
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package com.iluwatar.dirtyflag

// ABOUTME: A middle-layer component that calls DataFetcher to retrieve data from the back-end.
// ABOUTME: Caches the last fetched countries list and returns it when data has not changed.

/** A middle-layer app that calls/passes along data from the back-end. */
class World {

    private var countries: List<String> = emptyList()
    private val df = DataFetcher()

    /**
     * Calls [DataFetcher] to fetch data from back-end.
     *
     * @return List of strings
     */
    fun fetch(): List<String> {
        val data = df.fetch()
        countries = data.ifEmpty { countries }
        return countries
    }
}
